sympy/sets/sets.py

```diff
@@ -1336,23 +1310,30 @@ def _contains(self, other):
         return And(*[set.contains(other) for set in self.args])
 
     def __iter__(self):
-        no_iter = True
-        for s in self.args:
-            if s.is_iterable:
-                no_iter = False
-                other_sets = set(self.args) - set((s,))
-                other = Intersection(*other_sets, evaluate=False)
-                for x in s:
-                    c = sympify(other.contains(x))
-                    if c is S.true:
-                        yield x
-                    elif c is S.false:
-                        pass
-                    else:
-                        yield c
+        sets_sift = sift(self.args, lambda x: x.is_iterable)
+
+        completed = False
+        candidates = sets_sift[True] + sets_sift[None]
 
-        if no_iter:
-            raise ValueError("None of the constituent sets are iterable")
+        for s in candidates:
+            other_sets = set(self.args) - set((s,))
+            other = Intersection(*other_sets, evaluate=False)
+            completed = True
+            for x in s:
+                try:
+                    if x in other:
+                        yield x
+                except TypeError:
+                    completed = False
+            if completed:
+                raise StopIteration()
+
+        if not completed:
+            if not candidates:
+                raise ValueError("None of the constituent sets are iterable")
+            raise ValueError(
+                "The computation had not completed because of the "
```

Comment:
If the intended use here is that a caller would do `list(some_set)` but catch this exception then maybe it should be `NotIterableError` rather than `ValueError`. Built in Python types generally give `TypeError` when not iterable.

The second `ValueError` here could be something like `UndecidableMembershipError`. There are many places that could be used in sets. This exception needs to be caught since there is no flag like `is_iterable` to anticipate it.